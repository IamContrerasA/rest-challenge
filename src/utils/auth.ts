import jwt from 'jsonwebtoken'
import { PrismaClient, User } from '@prisma/client'
import { Request, Response, NextFunction } from 'express'
import bcrypt from 'bcrypt'

const prisma = new PrismaClient()
const JWT_SECRET = 'JWT_SECRET'

export const newToken = (user: User | null) => {
  return jwt.sign({ id: user?.id }, JWT_SECRET, {
    expiresIn: '100d',
  })
}

export const verifyToken = (token: string) =>
  new Promise((resolve, reject) => {
    jwt.verify(token, JWT_SECRET, (err, payload) => {
      if (err) return reject(err)
      resolve(payload)
    })
  })

export const signup = async (req: Request, res: Response): Promise<void> => {
  if (!req.body.email || !req.body.password) {
    res.status(400).send({ message: 'need email and password' })
  }

  try {
    let user: User
    bcrypt.hash(req.body.password, 8, async (err, hash) => {
      if (err) {
        throw new Error(`There is an error on hash password ${err}`)
      }
      user = await prisma.user.create({
        data: {
          email: req.body.email,
          password: hash,
        },
      })
    })

    const token = newToken(user!)
    res.status(201).send({ token })
  } catch (e) {
    return res.status(500).end()
  }
}

//checkpassword
function checkPassword(userPassord: string, password: string) {
  return new Promise((resolve, reject) => {
    bcrypt.compare(password, userPassord, (err, same) => {
      if (err) {
        console.log(err)
        return reject(err)
      }
      resolve(same)
    })
  })
}

export async function signin(req: Request, res: Response): Promise<void> {
  if (!req.body.email || !req.body.password) {
    res.status(400).send({ message: 'need email and password' })
  }

  const invalid = 'Invalid email and passoword combination'

  try {
    const user = await prisma.user.findUnique({
      where: { email: req.body.email },
    })

    if (!user) {
      res.status(401).send({ message: `${invalid}, no user found` })
    }

    const match = await checkPassword(user!.password, req.body.password)

    if (!match) {
      res.status(401).send({ message: `${invalid}, password dont match` })
    }

    const token = newToken(user)
    res.status(201).send({ token })
  } catch (e) {
    console.error(e)
    res.status(500).end()
  }
}

export const protect = async (
  req: Request,
  res: Response,
  next: NextFunction,
): Promise<void> => {
  const bearer = req.headers.authorization

  if (!bearer || !bearer.startsWith('Bearer ')) {
    return res.status(401).end()
  }

  const token = bearer.split('Bearer ')[1].trim()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let payload: any
  try {
    payload = await verifyToken(token)
  } catch (e) {
    return res.status(401).end()
  }

  const user = await prisma.user.findUnique({
    where: { id: payload.id },
    select: {
      id: true,
      firstname: true,
      surname: true,
      email: true,
      password: false,
    },
  })
  if (!user) {
    return res.status(401).end()
  }

  req.body.user = user
  next()
}
