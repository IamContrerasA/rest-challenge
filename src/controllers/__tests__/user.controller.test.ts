/* eslint-disable @typescript-eslint/no-explicit-any */
import { prisma, signup } from '../../utils/auth'
import {
  deleteUser,
  findUser,
  findUsers,
  publicEmail,
  publicName,
  updateUser,
} from '../users.controller'

let token: string
// test('token', async () => {
//   console.log(token)
// })
beforeAll(async () => {
  const req: any = {
    body: {
      email: 'myemail13@email.com',
      password: 'password13',
    },
  }

  const res: any = {
    status: function (statusNumber: number) {
      expect(statusNumber).toBe(201)
      return this
    },
    send: function (sendResult: any) {
      token = sendResult.token
      return this
    },
  }
  await signup(req, res)
})

afterAll(async () => {
  await prisma.$executeRaw(
    `TRUNCATE TABLE public."Comment" RESTART IDENTITY CASCADE;`,
  )
  await prisma.$executeRaw(
    `TRUNCATE TABLE public."Post" RESTART IDENTITY CASCADE;`,
  )
  await prisma.$executeRaw(
    `TRUNCATE TABLE public."User" RESTART IDENTITY CASCADE;`,
  )
  await prisma.$disconnect()
})

describe('test user controller happy way', () => {
  test('find all users', async () => {
    await new Promise((r) => setTimeout(r, 2000))
    const req: any = {}

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.users[0].email).toBe('myemail13@email.com')
        return this
      },
    }
    await findUsers(req, res)
  })

  test('find specific user', async () => {
    const req: any = { params: { id: 1 } }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.user.id).toBe(1)
        return this
      },
      end: () => this,
    }
    await findUser(req, res)
  })

  test('update specific user', async () => {
    const req: any = {
      params: { id: 1 },
      body: { firstname: 'nombre', user: { id: 1 } },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.user.id).toBe(1)
        return this
      },
      end: () => this,
    }
    await updateUser(req, res)
  })

  test('set public name to specific user', async () => {
    const req: any = {
      params: { id: 1 },
      body: { isNamePublic: false, user: { id: 1 } },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.user.isNamePublic).toBe(true)
        return this
      },
      end: () => this,
    }
    await publicName(req, res)
  })

  test('set public email to specific user', async () => {
    const req: any = {
      params: { id: 1 },
      body: { isEmailPublic: false, user: { id: 1 } },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.user.isEmailPublic).toBe(true)
        return this
      },
      end: () => this,
    }
    await publicEmail(req, res)
  })

  test('delete specific user', async () => {
    const req: any = {
      params: { id: 1 },
      body: { user: { id: 1 } },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.user.id).toBe(1)
        return this
      },
      end: () => this,
    }
    await deleteUser(req, res)
  })
})

describe('test user controller happy way', () => {
  test('find all users without users is invalid', async () => {
    const req: any = {}

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(404)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await findUsers(req, res)
  })

  test('find empty user is invalid', async () => {
    const req: any = { params: { id: 100 } }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(404)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await findUser(req, res)
  })

  test('update empty user is invalid', async () => {
    const req: any = {
      params: { id: 100 },
      body: { firstname: 'nombre', user: { id: 100 } },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(403)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await updateUser(req, res)
  })

  test('set public name to empty user is invalid', async () => {
    const req: any = {
      params: { id: 100 },
      body: { isNamePublic: false, user: { id: 100 } },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(403)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await publicName(req, res)
  })

  test('set public email to empty user is invalid', async () => {
    const req: any = {
      params: { id: 100 },
      body: { isEmailPublic: false, user: { id: 100 } },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(403)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await publicEmail(req, res)
  })

  test('delete empty user is invalid', async () => {
    const req: any = {
      params: { id: 1 },
      body: { user: { id: 1 } },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(403)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await deleteUser(req, res)
  })
})
