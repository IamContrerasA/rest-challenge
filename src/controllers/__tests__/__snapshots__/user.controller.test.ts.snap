// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test user controller error way delete empty user is invalid 1`] = `"Current user cannot edit the information of another user, only that of him"`;

exports[`test user controller error way find all users without users is invalid 1`] = `"Couldn't find any users jet"`;

exports[`test user controller error way find empty user is invalid 1`] = `"Couldn't find User with id = '100'"`;

exports[`test user controller error way set public email to empty user is invalid 1`] = `"Current user cannot edit the information of another user, only that of him"`;

exports[`test user controller error way set public name to empty user is invalid 1`] = `"Current user cannot edit the information of another user, only that of him"`;

exports[`test user controller error way update empty user is invalid 1`] = `"Current user cannot edit the information of another user, only that of him"`;
